lib/puppet/functions/join.rb

```diff
@@ -0,0 +1,99 @@
+# Joins an array of values into a string with elements separated by a delimiter.
+#
+# Supports up to three arguments
+# * first argument is required and must be an `Array`, or an `Iterable`
+# * second arguments is the delimiter, must be a `String` if given, and defaults to an empty string.
+# * third argument is a Boolean, when if true will also format nested hashes
+#
+# @example Typical use of `join`
+#
+# ```puppet
+# join(['a','b','c'], ",")
+# # Would result in: "a,b,c"
+# ```
+#
+# Details:
+# * The array is flattened before elements are joined, but flattening does not extend to arrays nested in hashes or other objects unless
+#   the third argument is `true`
+# * When an `Iterable` is used (such as the result from `reverse_each`) it is converted to an array before joining. This
```

Comment:
Isn't this just an implementation detail? Why is it relevant to the user that an `Iterable` is first converted to an `Array` (which in turn is another form of an `Iterable`)?